google-cloud-apis/src/com/google/cloud/tools/intellij/cloudapis/AddCloudLibrariesAction.java

```diff
@@ -48,52 +36,13 @@ public AddCloudLibrariesAction() {
   @Override
   @SuppressWarnings("MissingCasesInEnumSwitch")
   public void update(AnActionEvent e) {
-    boolean addLibrariesEnabled = false;
-    if (e.getProject() != null) {
-      Set<CloudLibrariesModuleSupportType> moduleSupportTypes =
-          Stream.of(ModuleManager.getInstance(e.getProject()).getModules())
-              .map(this::checkModuleForAddCloudLibraries)
-              .collect(Collectors.toSet());
-
-      addLibrariesEnabled = moduleSupportTypes.contains(CloudLibrariesModuleSupportType.SUPPORTED);
-      if (!addLibrariesEnabled) {
-        // update message to hint what is missing.
-        for (CloudLibrariesModuleSupportType supportType : moduleSupportTypes) {
-          switch (supportType) {
-            case MAVEN_REQUIRED:
-              e.getPresentation()
-                  .setDescription(
-                      GoogleCloudApisMessageBundle.message(
-                          "cloud.libraries.menu.action.maven.required.description"));
-              e.getPresentation()
-                  .setText(
-                      GoogleCloudApisMessageBundle.message(
-                          "cloud.libraries.menu.action.disabled.maven.text"));
-              break;
-            case APPENGINE_JAVA8_REQUIRED:
-              e.getPresentation()
-                  .setDescription(
-                      GoogleCloudApisMessageBundle.message(
-                          "cloud.libraries.menu.action.gae.java8.required.description"));
-              e.getPresentation()
-                  .setText(
-                      GoogleCloudApisMessageBundle.message(
-                          "cloud.libraries.menu.action.disabled.java8.text"));
-              break;
-          }
-        }
-      } else {
-        // standard message for a supported module action.
-        e.getPresentation()
-            .setDescription(
-                GoogleCloudApisMessageBundle.message("cloud.libraries.menu.action.description"));
-        e.getPresentation()
-            .setText(
-                GoogleCloudApisMessageBundle.message("cloud.libraries.menu.action.active.text"));
+    // iterate all available decorators and stop on first that decorated this action.
+    for (CloudApiActionDecoratorExtension decorator :
+        CloudApiActionDecoratorExtension.EP_NAME.getExtensions()) {
+      if (decorator.decorate(e)) {
```

Comment:
could we have a test to cover this mechanism? for example, that the correct decorator gets applied when multiple are registered etc.

Reply:
Always a good idea! I'll add a few, missed this.

Reply:
I added a unit test which is not quite simple since I think we didn't test using actual extension points before so I investigated easiest way to do it for now and then we can make this part of out unit test rule (`CloudToolsRule`) and potentially re-use for container tools testing as well.